giraph-core/src/main/java/org/apache/giraph/writable/kryo/GiraphClassResolver.java

```diff
@@ -0,0 +1,308 @@
+              true);
+    } catch (KeeperException e) {
+      throw new IllegalStateException(
+              "Failed to refresh kryo cache " +
+                      KRYO_REGISTERED_CLASS_PATH, e);
+    } catch (InterruptedException e) {
+      throw new IllegalStateException(
+              "Interrupted while refreshing kryo cache " +
+                      KRYO_REGISTERED_CLASS_PATH, e);
+    }
+
+    List<String> registeredList;
+    try {
+      registeredList =
+              ZK.getChildrenExt(KRYO_REGISTERED_CLASS_PATH,
+                      false,
+                      true,
+                      false);
+    } catch (KeeperException e) {
+      throw new IllegalStateException(
+        "Failed to retrieve child nodes for " + KRYO_REGISTERED_CLASS_PATH, e);
+    } catch (InterruptedException e) {
+      throw new IllegalStateException(
+        "Interrupted while retrieving child nodes for " +
+                KRYO_REGISTERED_CLASS_PATH, e);
+    }
+
+    for (String name : registeredList) {
+      String className = name.substring(0,
+              name.length() - SEQUENCE_NUMBER_LENGTH);
+      int classId = Integer.parseInt(
+              name.substring(name.length() - SEQUENCE_NUMBER_LENGTH));
+
+      if (MIN_CLASS_ID == -1) {
+        MIN_CLASS_ID = classId;
+      }
+
+      int adjustedId = classId - MIN_CLASS_ID + BASE_CLASS_ID;
+      if (CLASS_NAME_TO_ID.putIfAbsent(className, adjustedId) == null) {
+        ID_TO_CLASS_NAME.put(adjustedId, className);
+      }
+    }
+  }
+
+  /**
+   * Gets ID for the given class name.
+   * @param className Class name
+   * @return class id Class ID
+   */
+  public static int getClassId(String className) {
+    if (CLASS_NAME_TO_ID.containsKey(className)) {
+      return CLASS_NAME_TO_ID.get(className);
+    }
+    synchronized (GiraphClassResolver.class) {
+      if (CLASS_NAME_TO_ID.containsKey(className)) {
+        return CLASS_NAME_TO_ID.get(className);
+      }
+      refreshCache();
+
+      if (!CLASS_NAME_TO_ID.containsKey(className)) {
+        createClassName(className);
+        refreshCache();
+      }
+    }
+
+    if (!CLASS_NAME_TO_ID.containsKey(className)) {
+      throw new IllegalStateException("Failed to assigned id to " + className);
+    }
+
+    return CLASS_NAME_TO_ID.get(className);
+  }
+
+  /**
+   * Get class name for given ID.
+   * @param id class ID
+   * @return class name
+   */
+  public static String getClassName(int id) {
+    if (ID_TO_CLASS_NAME.containsKey(id)) {
+      return ID_TO_CLASS_NAME.get(id);
+    }
+    synchronized (GiraphClassResolver.class) {
+      if (ID_TO_CLASS_NAME.containsKey(id)) {
+        return ID_TO_CLASS_NAME.get(id);
+      }
+      refreshCache();
+    }
+
+    if (!ID_TO_CLASS_NAME.containsKey(id)) {
+      throw new IllegalStateException("ID " + id + " doesn't exist");
+    }
+    return ID_TO_CLASS_NAME.get(id);
+  }
+
+  @Override
+  public Registration register(Registration registration) {
```

Comment:
The default implementation also handles the case when the class is non-explicitly registered, i.e. it is not assigned to an an integer. Since in this implementation all classes are assigned to integers agreed by all kryo instances, they all treated like explicitly assigned classes. This implementation throws an exception if it encounters a class that has not been assigned to an integer yet.